sw/device/tests/dif/dif_otbn_sanitytest.c

```diff
@@ -0,0 +1,271 @@
+  CHECK(dif_otbn_start(&otbn, start_address) == kDifOtbnOk);
+}
+
+/**
+ * Busy wait for OTBN to be done with its operation.
+ */
+static void otbn_wait_for_done() {
+  bool busy = true;
+  while (busy) {
+    CHECK(dif_otbn_is_busy(&otbn, &busy) == kDifOtbnOk,
+          "Unable to get busy status from OTBN");
+  }
+}
+
+/**
+ * Initialize OTBN's data memory with zeros
+ */
+static void zero_dmem() {
+  int dmem_size_words = dif_otbn_get_dmem_size_bytes(&otbn) / sizeof(uint32_t);
+  for (int i = 0; i < dmem_size_words; ++i) {
+    const uint32_t zero = 0;
+    dif_otbn_result_t rv =
+        dif_otbn_dmem_write(&otbn, i * sizeof(uint32_t), &zero, sizeof(zero));
+    CHECK(rv == kDifOtbnOk, "Error zeroing word %d in OTBN DMEM: %d", i, rv);
+  }
+}
+
+/**
+ * Run a 384-bit Barrett Multiplication on OTBN and check its result.
+ *
+ * This test is not aiming to exhaustively test the Barrett multiplication
+ * itself, but test the interaction between device software and OTBN. As such,
+ * only trivial parameters are used.
+ *
+ * The code executed on OTBN can be found in sw/otbn/code-snippets/barrett384.s.
+ * The entry point wrap_barrett384() is called according to the calling
+ * convention described in the OTBN assembly code file.
+ */
+static void test_barrett384() {
+  enum { kDataSizeBytes = 48 };
+
+  zero_dmem();
+
+  load_app(&kOtbnAppBarrett384);
+
+  // a, first operand
+  static const uint8_t a[kDataSizeBytes] = {10};
+
+  // b, second operand
+  static uint8_t b[kDataSizeBytes] = {20};
+
+  // m, modulus, max. length 384 bit with 2^384 > m > 2^383
+  // We choose the modulus of P-384: m = 2**384 - 2**128 - 2**96 + 2**32 - 1
+  static const uint8_t m[kDataSizeBytes] = {
+      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe, 0xff, 0xff, 0xff, 0xff,
+      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff};
+
+  // u, pre-computed Barrett constant (without u[384]/MSb of u which is always 1
+  // for the allowed range but has to be set to 0 here).
+  // u has to be pre-calculated as u = floor(2^768/m).
+  static const uint8_t u[kDataSizeBytes] = {
+      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
+      0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x01};
+
+  // c, result, max. length 384 bit.
+  uint8_t c[kDataSizeBytes] = {0};
+
+  // c = (a * b) % m = (10 * 20) % m = 200
+  static const uint8_t c_expected[kDataSizeBytes] = {200};
+
+  LOG_INFO("Writing input arguments to DMEM");
+  // TODO: Use symbols from the application to load these parameters once they
+  // are available (#3998).
+  CHECK(dif_otbn_dmem_write(&otbn, /*offset_bytes=*/0, (uint32_t *)&a,
```

Comment:
These casts shouldn't be necessary, right?

(I accidentally mentioned this in another PR that contained this code.)

Reply:
They are no longer necessary, I removed them.